cmd/clusterctl/pkg/client/cluster/move.go

```diff
@@ -202,11 +232,412 @@ func (o *objectGraph) Discovery(namespace string, types []metav1.TypeMeta) error
+	return o.deleteTreeFromSource(cluster)
+}
+
+const (
+	retryDeleteSourceObject         = 3
+	retryIntervalDeleteSourceObject = 1 * time.Second
+)
+
+// deleteTreeFromSource deletes a node and all the dependent nodes from the source management cluster.
+func (o *objectGraph) deleteTreeFromSource(nodeToDelete *node) error {
+
+	// Deletes - in parallel - the dependents nodes and the softDependents nodes (with the respective object tree).
+	// The approach to concurrency here comes from golang.org/x/sync/errgroup.
+	var wg sync.WaitGroup
+	var gerr error
+	var errOnce sync.Once
+
+	for dependent := range nodeToDelete.dependents {
+		wg.Add(1)
+		go func(dependent *node) {
+			defer wg.Done()
+			if err := o.deleteTreeFromSource(dependent); err != nil {
+				errOnce.Do(func() { gerr = err })
+			}
+		}(dependent)
+	}
+
+	for dependent := range nodeToDelete.softDependents {
+		wg.Add(1)
+		go func(dependent *node) {
+			defer wg.Done()
+			if err := o.deleteTreeFromSource(dependent); err != nil {
+				errOnce.Do(func() { gerr = err })
+			}
+		}(dependent)
+	}
+
+	wg.Wait()
+	if gerr != nil {
+		return gerr
+	}
+
+	// Delete the Kubernetes object corresponding to the current node.
+	// Nb. The operation is wrapped in a retry loop to make move more resilient to unexpected conditions.
+	err := retry(retryDeleteSourceObject, retryIntervalDeleteSourceObject, o.log, func() error {
+		return o.deleteSourceObject(nodeToDelete)
+	})
+
+	return err
+}
+
+// deleteSourceObject deletes the object corresponding to the node from the source management cluster, taking care of removing all the finalizers so
+// the objects gets immediately deleted (force delete).
+func (o *objectGraph) deleteSourceObject(nodeToDelete *node) error {
+	o.log.V(1).Info("Deleting", nodeToDelete.identity.Kind, nodeToDelete.identity.Name, "Namespace", nodeToDelete.identity.Namespace)
+
+	cFrom, err := o.fromProxy.NewClient()
+	if err != nil {
+		return err
+	}
+
+	// Get the source object
+	sourceObj := &unstructured.Unstructured{}
+	sourceObj.SetAPIVersion(nodeToDelete.identity.APIVersion)
+	sourceObj.SetKind(nodeToDelete.identity.Kind)
+	sourceObjKey := client.ObjectKey{
+		Namespace: nodeToDelete.identity.Namespace,
+		Name:      nodeToDelete.identity.Name,
+	}
+
+	if err := cFrom.Get(ctx, sourceObjKey, sourceObj); err != nil {
+		if apierrors.IsNotFound(err) {
+			//If the object is already deleted, move on.
+			o.log.V(2).Info("Object already deleted, skipping delete for", nodeToDelete.identity.Kind, nodeToDelete.identity.Name, "Namespace", nodeToDelete.identity.Namespace)
+			return nil
+		}
+		return errors.Wrapf(err, "error reading %q %s/%s",
+			sourceObj.GroupVersionKind(), sourceObj.GetNamespace(), sourceObj.GetName())
+	}
+
+	if len(sourceObj.GetFinalizers()) > 0 {
+		sourceObjWithoutFinalizers := sourceObj.DeepCopy()
+		sourceObjWithoutFinalizers.SetFinalizers([]string{})
+
+		resourcePatch := client.MergeFrom(sourceObj)
```

Comment:
Can you hard-code the patch bytes for this?

Reply:
Also, would sourceObj ever be a namespace?

Reply:
No, it can be a Namespace (only Secret, ConfigMap or one of the Kind defined in the CRDs).

Reply:
Patch hard-coded